Resolve traceback error when using variables in template triggers

ehendrix23 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Fix traceback error ```NoneType can't be used in await expression``` when using variables in template triggers.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #72434
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ehendrix23 committed:
Change callback to async

ehendrix23 commented:
Just doing a quick check on status. I have not seen these errors occur at all within my installation after implementing this fix and everything has been working fine related to it.

emontnemery commented:
The problem is caused by a regression introduced by https://github.com/home-assistant/core/pull/68275
While this PR hides the problem in the `template` integration, the same problem exists also in `websocket_api` etc.

Instead of this, we should fix the `_trigger_action_wrapper` which was introduced by `https://github.com/home-assistant/core/pull/68275` to handle either regular functions or coroutine functions.

ehendrix23 committed:
Revert initial fix

ehendrix23 committed:
Handle coroutine and callback

ehendrix23 commented:
I've reverted the change I initially did within the template and updated ```_trigger_action_wrapper``` instead using the example provided.
Ran test on everything under components/template and everything under helpers/test_trigger without issues.
Updated my local copy for home assistant and no errors, also no errors as reported originally.

Let me know if these changes look good or further updates required. :-)

emontnemery committed:
Don't return a coroutine function when wrapping a function

emontnemery committed:
Add test

emontnemery commented:
Thanks @ehendrix23 !

Unfortunately, the more complicated solution I linked to is needed to not return a coroutine function when wrapping a regular function. I updated the PR accordingly + added a test.

balloob commented on homeassistant/helpers/trigger.py:
```diff
@@ -101,20 +107,51 @@ async def async_validate_trigger_config(
-        trigger_variables = conf[CONF_VARIABLES]
-        run_variables.update(trigger_variables.async_render(hass, run_variables))
-        await action(run_variables, context)
+    # Check for partials to properly determine if coroutine function
+    check_func = action
+    while isinstance(check_func, functools.partial):
+        check_func = check_func.func
```
Not needed if you use [`inspect.iscoroutinefunction`](https://docs.python.org/3/library/inspect.html#inspect.iscoroutinefunction)

emontnemery replied:
While it's not needed for `inspect.iscoroutinefunction`, it's needed for `is_callback`.
If we teach `is_callback` about unwinding partials we can indeed drop it, that's for another PR though.

ehendrix23 commented:
Thanks. I didn't think the more complex one was required but I guess I was wrong. :-)

Tested on my local instance and issue is not occurring. So looks like this is working.

Saw the comment from balloob. And you acknowledging but not sure if anything further is required as it looks like there was a force push but I don't see any difference in that section. :-)

emontnemery commented:
@ehendrix23 right, I first implemented @balloob's suggestion, then I realized we need the unwinding also for `is_partial` so I reverted it. Sorry about the force push.
